src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java

```diff
@@ -302,34 +304,45 @@ private boolean runRepair() {
 
       LazyInitializer<Set<String>> busyHosts = new BusyHostsInitializer(cluster);
 
-      // If we're using a distributed storage, we need to synchronize with other Reaper instances
-      // So we don't start too many segments at the same time
-      if (!lockSegmentRunners()) {
-        LOG.debug(
-            "Not allowed to run the segment for now as another Reaper holds the lock for repair run {}. "
-            + "Will try again later",
-            segment.getRunId());
-        return false;
-      }
       if (!canRepair(segment, keyspace, coordinator, cluster, busyHosts)) {
         LOG.info(
-            "Cannot run segment {} for repair {} at the moment. Will try again later",
-            segmentId,
-            segment.getRunId());
-        SEGMENT_RUNNERS.remove(segment.getId());
+            "Cannot run segment {} for repair {} at the moment. Will try again later", segmentId, segment.getRunId());
+
         try {
           Thread.sleep(SLEEP_TIME_AFTER_POSTPONE_IN_MS);
-        } catch (InterruptedException e) {
-          LOG.debug("Interrupted while sleeping after a segment was postponed... weird stuff...");
-        }
-        releaseSegmentRunners();
+        } catch (InterruptedException ignore) { }
         return false;
       }
 
       try (Timer.Context cxt1 = context.metricRegistry.timer(metricNameForRepairing(segment)).time()) {
         try {
           LOG.debug("Enter synchronized section with segment ID {}", segmentId);
           synchronized (condition) {
+            if (!lockSegmentRunners()) {
+              // XXX – not expected to happen, STARTED run state should be "good" (opportunistic) enough
+              LOG.warn(
+                  "Cannot run segment {} as another Reaper holds the lock on repair run {}. Will try again later",
+                  segmentId,
+                  segment.getRunId());
+
+              return false;
+            }
+
+            // double-locking-idiom, only applies to non-incremental and distributed storage
+            if (!repairUnit.getIncrementalRepair()
+                && context.storage instanceof IDistributedStorage
+                && !canRepair(segment, keyspace, coordinator, cluster, busyHosts)) {
```

Comment:
does the jmx metrics have to happen inside the lock? if not we can simplify the second `canRepair(..)` (this one).

Reply:
possibility to do `assert nodesReadyForNewRepair (..)` here…